### PR TITLE
[DirectX] Add support to lower LLVM intrinsics ceil, cos, fabs, floor and smax to DXIL Ops.

### PR DESCRIPTION
**Pull request by bharadwajy**
Also add a DXILStrengthReduce pass to facilitate rewriting LLVM IR in preparation for DXIL lowering. Moved rewriting of FNeg from DXILModulePrepare to this pass and added support to rewrite abs intrinsic to the pass.

Add tests for each of the instructions being newly lowered and rewritten.

### CODE REVIEW

**Comment by bharadwajy** [comment]
@llvm-beanz @bogner

File: llvm/lib/Target/DirectX/DirectX.h [hunk 1]
```diff
@@ -28,6 +28,12 @@ void initializeDXILPrepareModulePass(PassRegistry &);
 /// Pass to convert modules into DXIL-compatable modules
 ModulePass *createDXILPrepareModulePass();
 
+/// Initializer for DXIL strength reduce
+void initializeDXILStrengthReducePass(PassRegistry &);
+
+/// Pass to reduce strength during lowering into DXIL-compatable modules
+ModulePass *createDXILStrengthReducePass();
```
Comment (python3kgae) [inline llvm/lib/Target/DirectX/DirectX.h:35]: Do we need DXILStrengthReducePass for add these new intrinsics?

Comment (bharadwajy) [reply llvm/lib/Target/DirectX/DirectX.h:35]: This pass is introduced to rewrite intrinsics such as `llvm.abs.*` (integer abs) with no corresponding DXIL intrinsic (AFAIK). For example, `llvm.abs` is rewritten in this pass as a combination of `sub` (LLVM instruction) and `smax` (LLVM intrinsic) - as done in DXC.

**Comment by bharadwajy** [comment]
I think this PR needs some refinements. Marking it as draft to get feedback, if possible.

File: llvm/lib/Target/DirectX/DirectX.h (hunk 1 above)
Comment (dmpots) [reply llvm/lib/Target/DirectX/DirectX.h:35]: It looks like you are only implementing the strenght reduce for `llvm.abs.*` and `fneg` here, but are adding many other intrinsics.

I think that in general we should strive to make the PRs in the smallest self-contained way possible. This makes it easier to review (and people are also usually more willing to jump in and review smaller changes).

I see this as three logical PRs here

1. Add the strength reduce pass and handle the `fneg` instruction.
2. Extend the strength reduce pass to also lower the `llvm.abs` intrinsics
3. Add the other intrinsics that do not depend on the strenght reduce pass (e.g. `llvm.cos`).

The first PR is basically a refactoring (do we have exisiting test coverage for `fneg`)? And separating the other PRs makes them easier to review and give constructive feedback.

File: llvm/lib/Target/DirectX/DXILStrengthReduce.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+public:
+  bool runOnModule(Module &M) override {
+    for (auto &F : make_early_inc_range(M.functions())) {
+      IRBuilder<> IRB(M.getContext());
+      // Reduce strength of LLVM intrinsics
+      // Flag to indicate if the intrinsic has been replaced. This ensures any
+      // other functions with no uses are not deleted in this pass.
+      bool IntrinsicReplaced = false;
+      if (F.isDeclaration()) {
```
Comment (dmpots) [inline llvm/lib/Target/DirectX/DXILStrengthReduce.cpp:43]: I find big chunks of inlined code hard to read. If we expect this pass to be extended in the future I think it is worth separating out some functions to make the overall logic easier to follow like

```
void strengthReduceAbs(Function &F) {
    // Strength reduce abs to max.
}

void strengthReduceFunctionDecl(Function &F) {
    switch (F.getIntrinsicID()) {
        case Intrinsic::abs: strengthReduceAbs(F); break;
    }
    // Handle FNeg for now.
}

void strengthReduceFunctionBody(Function &F)
{
    // Handle FNeg for now.
}


runOnModule() {
    for (auto &F : make_early_inc_range(M.functions())) {
        if (F.isDeclaration())
            strengthReduceFunctionDecl(F);
        else
            strenghtReduceFunctionBody(F)
        }
}
```

File: llvm/test/CodeGen/DirectX/abs.ll [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+; RUN: opt -S -dxil-strength-reduce < %s | FileCheck %s -check-prefix=TEST_SR
+
+; Make sure output of strength reduction pass is lowered to DXIL code as expected.
+; RUN: opt -S -dxil-strength-reduce -dxil-op-lower < %s | FileCheck %s -check-prefix=TEST_SR_OP_LOWER
+
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
+target triple = "dxil-unknown-shadermodel6.5-compute"
+
+@"?a@@3HA" = local_unnamed_addr global i32 0, align 4
```
Comment (dmpots) [inline llvm/test/CodeGen/DirectX/abs.ll:11]: I find using the mangled names makes the tests harder to read. They are used throughout these tests cases for both function names and variables.

Using simpler names makes it easier to read I think.

File: llvm/test/CodeGen/DirectX/abs.ll [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  ret void
+}
+
+; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none)
+declare i32 @llvm.abs.i32(i32, i1 immarg) #2
+
+; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(readwrite, argmem: none, inaccessiblemem: none)
+define void @"?test_i64@@YAXI@Z"(i32 noundef %GI) local_unnamed_addr #1 {
+entry:
+  %0 = load i64, ptr @"?d@@3JA", align 8, !tbaa !8
```
Comment (dmpots) [inline llvm/test/CodeGen/DirectX/abs.ll:37]: These tests also have a lot of unrelated code to what we are actually trying to test.

I think an even simpler test like this would serve the same purpose.
```
define void @test_i64(i32 noundef %a)
{
    %elt.abs = tail call i64 @llvm.abs.i64(i64 %a, i1 false)
}
```

**Comment by bharadwajy** [comment]
Thanks for all the feedback. Will follow up as suggested.
